contributors/design-proposals/statefulset-update.md

```diff
@@ -0,0 +1,850 @@
+ 
+```yaml
+apiVersion: apps/v1beta1
+kind: StatefulSet
+metadata:
+  name: web
+spec:
+  serviceName: "nginx"
+  replicas: 3
+  template:
+    metadata:
+      labels:
+        app: nginx
+    spec:
+      containers:
+      - name: nginx
+        image: gcr.io/google_containers/nginx-slim:0.9
+        ports:
+        - containerPort: 80
+          name: web
+        volumeMounts:
+        - name: www
+          mountPath: /usr/share/nginx/html
+      revisionPartition: 2
+  volumeClaimTemplates:
+  - metadata:
+      name: www
+      annotations:
+        volume.alpha.kubernetes.io/storage-class: anything
+    spec:
+      accessModes: [ "ReadWriteOnce" ]
+      resources:
+        requests:
+          storage: 1Gi
+```
+
+Users can also simultaneously scale up and add a canary. This reduces risk 
+for some deployment scenarios by adding additional capacity for the canary. 
+
+```yaml
+apiVersion: apps/v1beta1
+kind: StatefulSet
+metadata:
+  name: web
+spec:
+  serviceName: "nginx"
+  replicas: 4
+  template:
+    metadata:
+      labels:
+        app: nginx
+    spec:
+      containers:
+      - name: nginx
+        image: gcr.io/google_containers/nginx-slim:0.9
+        ports:
+        - containerPort: 80
+          name: web
+        volumeMounts:
+        - name: www
+          mountPath: /usr/share/nginx/html
+      partitionOrdinal: 3
+  volumeClaimTemplates:
+  - metadata:
+      name: www
+      annotations:
+        volume.alpha.kubernetes.io/storage-class: anything
+    spec:
+      accessModes: [ "ReadWriteOnce" ]
+      resources:
+        requests:
+          storage: 1Gi
+```
+
+### Staged Roll Outs
+Users can create a canary using `kubectl apply`. The only difference between a
+ [canary](#canaries) and a staged roll out is that the `.Spec.RevisionPartition` 
+ is set to value less than `.Spec.Replicas - 1`.
+ 
+```yaml
+apiVersion: apps/v1beta1
+kind: StatefulSet
+metadata:
+  name: web
+spec:
+  serviceName: "nginx"
+  replicas: 3
+  template:
+    metadata:
+      labels:
+        app: nginx
+    spec:
+      containers:
+      - name: nginx
+        image: gcr.io/google_containers/nginx-slim:0.9
+        ports:
+        - containerPort: 80
+          name: web
+        volumeMounts:
+        - name: www
+          mountPath: /usr/share/nginx/html
+      revisionParition: 2
+  volumeClaimTemplates:
+  - metadata:
+      name: www
+      annotations:
+        volume.alpha.kubernetes.io/storage-class: anything
+    spec:
+      accessModes: [ "ReadWriteOnce" ]
+      resources:
+        requests:
+          storage: 1Gi
+```
+
+Staged roll outs can be used to roll out a configuration, image, or resource 
+update to some portion of the fleet maintained by the StatefulSet prior to 
+updating the entire fleet. It is useful to support linear, geometric, and 
+exponential roll out of an update. Users can modify the 
+`.Spec.RevisionPartition` to allow the roll out to progress.
```

Comment:
My thought is that if we have termination reason and local storage backed PVs (both future features) we probably need to support the ability to do staged roll out in this way. Mainly this would be to support operators that manage this type of worklow using StatefulSet as their primitive.